In [1]:
import pymysql
pymysql.install_as_MySQLdb()
%reload_ext sql

In [22]:
%sql mysql://mag129:Th2Usy9s@mqm-db/

'Connected: mag129@'

## TASK1 

In [3]:
%%sql

USE lahman2016

0 rows affected.


[]

In [17]:
%%sql
SELECT *
FROM(SELECT yearID,teamID,W/G AS Winning_Perc,
            RANK() OVER (Partition by yearID ORDER BY Winning_Perc desc) AS RANK
            FROM  Teams
WHERE yearID>=2013 AND yearID<=2016) AS hey
WHERE teamID='ARI';

4 rows affected.


[(2013, 'ARI', Decimal('0.5000'), 16),
 (2014, 'ARI', Decimal('0.3951'), 30),
 (2015, 'ARI', Decimal('0.4877'), 18),
 (2016, 'ARI', Decimal('0.4259'), 24)]

In [6]:
%%sql
SELECT yearID,teamID,RANK() OVER (Partition by yearID ORDER BY HR desc) as Hr_Rank,
       ROUND((H+BB+HBP)/(AB+BB+HBP+SF),2) AS onbase_percentage,
       RANK() OVER (Partition by yearID ORDER BY onbase_percentage desc) AS onbase_rank,
       RANK() OVER (Partition by yearID ORDER BY ERA ASC) as era_rank,
       RANK() OVER (Partition by yearID ORDER BY FP desc) as Fp_Rank, 
       LgWin
FROM   Teams
WHERE  yearID>=2000 AND yearID<=2016;

510 rows affected.


[(2000, 'ANA', 4, 0.35, 9, 21, 23, 'N'),
 (2000, 'ARI', 16, 0.33, 23, 7, 11, 'N'),
 (2000, 'ATL', 16, 0.35, 9, 1, 21, 'N'),
 (2000, 'BAL', 14, 0.34, 15, 27, 14, 'N'),
 (2000, 'BOS', 23, 0.34, 15, 5, 11, 'N'),
 (2000, 'CHA', 8, 0.36, 2, 14, 23, 'N'),
 (2000, 'CHN', 15, 0.33, 23, 25, 6, 'N'),
 (2000, 'CIN', 11, 0.34, 15, 6, 11, 'N'),
 (2000, 'CLE', 7, 0.37, 1, 18, 1, 'N'),
 (2000, 'COL', 25, 0.36, 2, 26, 2, 'N'),
 (2000, 'DET', 19, 0.34, 15, 15, 6, 'N'),
 (2000, 'FLO', 26, 0.33, 23, 12, 19, 'N'),
 (2000, 'HOU', 1, 0.36, 2, 28, 23, 'N'),
 (2000, 'KCA', 28, 0.35, 9, 29, 6, 'N'),
 (2000, 'LAN', 9, 0.34, 15, 2, 23, 'N'),
 (2000, 'MIL', 19, 0.33, 23, 13, 14, 'N'),
 (2000, 'MIN', 30, 0.34, 15, 23, 6, 'N'),
 (2000, 'MON', 18, 0.33, 23, 22, 23, 'N'),
 (2000, 'NYA', 10, 0.35, 9, 16, 14, 'Y'),
 (2000, 'NYN', 12, 0.35, 9, 3, 19, 'Y'),
 (2000, 'OAK', 3, 0.36, 2, 11, 23, 'N'),
 (2000, 'PHI', 29, 0.33, 23, 17, 6, 'N'),
 (2000, 'PIT', 22, 0.34, 15, 20, 21, 'N'),
 (2000, 'SDN', 27, 0.33, 23, 10, 30, 'N'),
 (2000, 'SEA', 12, 0.36, 2, 9, 4, 'N'),
 (2000, 'SFN', 6, 0.36, 2, 4, 2, 'N'),
 (2000, 'SLN', 5, 0.36, 2, 8, 14, 'N'),
 (2000, 'TBA', 24, 0.33, 23, 19, 14, 'N'),
 (2000, 'TEX', 21, 0.35, 9, 30, 23, 'N'),
 (2000, 'TOR', 2, 0.34, 15, 23, 4, 'N'),
 (2001, 'ANA', 24, 0.33, 13, 12, 6, 'N'),
 (2001, 'ARI', 7, 0.34, 6, 4, 1, 'Y'),
 (2001, 'ATL', 17, 0.32, 20, 2, 6, 'N'),
 (2001, 'BAL', 28, 0.32, 20, 22, 25, 'N'),
 (2001, 'BOS', 13, 0.33, 13, 9, 18, 'N'),
 (2001, 'CHA', 3, 0.33, 13, 19, 18, 'N'),
 (2001, 'CHN', 15, 0.34, 6, 7, 11, 'N'),
 (2001, 'CIN', 16, 0.32, 20, 24, 27, 'N'),
 (2001, 'CLE', 5, 0.35, 2, 20, 11, 'N'),
 (2001, 'COL', 4, 0.35, 2, 29, 5, 'N'),
 (2001, 'DET', 27, 0.32, 20, 27, 25, 'N'),
 (2001, 'FLO', 19, 0.33, 13, 15, 6, 'N'),
 (2001, 'HOU', 7, 0.35, 2, 16, 11, 'N'),
 (2001, 'KCA', 25, 0.32, 20, 25, 18, 'N'),
 (2001, 'LAN', 9, 0.32, 20, 13, 18, 'N'),
 (2001, 'MIL', 6, 0.32, 20, 20, 6, 'N'),
 (2001, 'MIN', 20, 0.34, 6, 17, 11, 'N'),
 (2001, 'MON', 29, 0.32, 20, 23, 11, 'N'),
 (2001, 'NYA', 10, 0.33, 13, 6, 11, 'Y'),
 (2001, 'NYN', 26, 0.32, 20, 8, 6, 'N'),
 (2001, 'OAK', 11, 0.35, 2, 2, 24, 'N'),
 (2001, 'PHI', 20, 0.33, 13, 9, 3, 'N'),
 (2001, 'PIT', 22, 0.31, 30, 28, 27, 'N'),
 (2001, 'SDN', 22, 0.34, 6, 18, 30, 'N'),
 (2001, 'SEA', 18, 0.36, 1, 1, 1, 'N'),
 (2001, 'SFN', 2, 0.34, 6, 11, 18, 'N'),
 (2001, 'SLN', 11, 0.34, 6, 5, 11, 'N'),
 (2001, 'TBA', 30, 0.32, 20, 26, 29, 'N'),
 (2001, 'TEX', 1, 0.34, 6, 30, 18, 'N'),
 (2001, 'TOR', 14, 0.33, 13, 14, 3, 'N'),
 (2002, 'ANA', 21, 0.34, 5, 4, 2, 'Y'),
 (2002, 'ARI', 14, 0.35, 1, 10, 5, 'N'),
 (2002, 'ATL', 17, 0.33, 15, 1, 17, 'N'),
 (2002, 'BAL', 14, 0.31, 28, 20, 5, 'N'),
 (2002, 'BOS', 9, 0.35, 1, 7, 13, 'N'),
 (2002, 'CHA', 3, 0.34, 5, 21, 10, 'N'),
 (2002, 'CHN', 5, 0.32, 20, 18, 21, 'N'),
 (2002, 'CIN', 11, 0.33, 15, 17, 21, 'N'),
 (2002, 'CLE', 7, 0.32, 20, 25, 21, 'N'),
 (2002, 'COL', 21, 0.34, 5, 28, 17, 'N'),
 (2002, 'DET', 30, 0.3, 30, 26, 29, 'N'),
 (2002, 'FLO', 24, 0.34, 5, 19, 13, 'N'),
 (2002, 'HOU', 12, 0.34, 5, 12, 2, 'N'),
 (2002, 'KCA', 26, 0.32, 20, 29, 24, 'N'),
 (2002, 'LAN', 20, 0.32, 20, 4, 5, 'N'),
 (2002, 'MIL', 27, 0.32, 20, 23, 13, 'N'),
 (2002, 'MIN', 12, 0.33, 15, 14, 1, 'N'),
 (2002, 'MON', 18, 0.33, 15, 11, 28, 'N'),
 (2002, 'NYA', 2, 0.35, 1, 8, 24, 'N'),
 (2002, 'NYN', 19, 0.32, 20, 9, 30, 'N'),
 (2002, 'OAK', 4, 0.34, 5, 3, 10, 'N'),
 (2002, 'PHI', 14, 0.34, 5, 15, 2, 'N'),
 (2002, 'PIT', 25, 0.32, 20, 16, 17, 'N'),
 (2002, 'SDN', 28, 0.32, 20, 22, 24, 'N'),
 (2002, 'SEA', 21, 0.35, 1, 13, 5, 'N'),
 (2002, 'SFN', 6, 0.34, 5, 2, 5, 'Y'),
 (2002, 'SLN', 10, 0.34, 5, 6, 13, 'N'),
 (2002, 'TBA', 29, 0.31, 28, 30, 24, 'N'),
 (2002, 'TEX', 1, 0.34, 5, 27, 10, 'N'),
 (2002, 'TOR', 8, 0.33, 15, 24, 17, 'N'),
 (2003, 'ANA', 24, 0.33, 14, 15, 17, 'N'),
 (2003, 'ARI', 22, 0.33, 14, 6, 10, 'N'),
 (2003, 'ATL', 3, 0.35, 3, 12, 20, 'N'),
 (2003, 'BAL', 22, 0.32, 23, 22, 10, 'N'),
 (2003, 'BOS', 2, 0.36, 1, 17, 17, 'N'),
 (2003, 'CHA', 5, 0.33, 14, 13, 8, 'N'),
 (20

In [12]:
%%sql
# only displaying Cham and NonCham
With t1 as 

(

SELECT yearID,
       teamID,
       RANK() OVER (Partition by yearID ORDER BY HR desc) as Hr_Rank,
       ROUND((H+BB+HBP)/(AB+BB+HBP+SF),2) AS onbase_percentage,
       RANK() OVER (Partition by yearID ORDER BY onbase_percentage desc) AS onbase_rank,
       RANK() OVER (Partition by yearID ORDER BY ERA ASC) as era_rank,
       RANK() OVER (Partition by yearID ORDER BY FP desc) as Fp_Rank, LgWin
FROM  Teams
WHERE yearID>=2000 AND yearID<=2016)

SELECT *
FROM(SELECT LgWin,AVG(Hr_Rank) OVER(Partition BY LgWin),AVG(onbase_rank) OVER(Partition BY LgWin),AVG(era_rank) OVER(Partition BY LgWin),AVG(Fp_Rank) OVER(Partition BY LgWin)
FROM t1
WHERE yearID>=2000 AND yearID<=2016) AS HEY
GROUP BY LgWin;

2 rows affected.


[('N', Decimal('15.6092'), Decimal('12.4769'), Decimal('15.9895'), Decimal('14.0714')),
 ('Y', Decimal('12.0294'), Decimal('7.7941'), Decimal('7.2647'), Decimal('10.4118'))]

In [14]:
%%sql
SELECT teamID,
       AVG(Hr_Rank),
       AVG(onbase_rank),
       AVG(era_rank), 
       AVG(Fp_Rank)
FROM(
       SELECT yearID,
              teamID,
              RANK() OVER (Partition by yearID ORDER BY HR desc) as Hr_Rank,
              ROUND((H+BB+HBP)/(AB+BB+HBP+SF),2) AS onbase_percentage,
              RANK() OVER (Partition by yearID ORDER BY onbase_percentage desc) AS onbase_rank,
              RANK() OVER (Partition by yearID ORDER BY ERA ASC) as era_rank,
              RANK() OVER (Partition by yearID ORDER BY FP desc) as Fp_Rank
FROM  Teams
WHERE yearID>=2013 AND yearID<=2016) AS babasss
WHERE teamID='ARI';

1 rows affected.


[('ARI', Decimal('20.7500'), Decimal('13.2500'), Decimal('22.5000'), Decimal('11.5000'))]

#Which statistic(s) better differentiates league champions from all other teams?  
- ERA_Rank differentiate league champions from all other teams.
Where is your team currently doing well?  
- ARI has higher ranks in the HR_Rank, Onbase_Rank, and Fp_Rank
Are there potential opportunities for improvement?
- ARI could improve on the ERA_Rank

In [16]:
%%sql
With t1
as
(
SELECT *,RANK() OVER (Partition by POS,yearID ORDER BY InnOuts desc) as siralama
FROM Fielding
WHERE InnOuts != '' AND yearID>=2013 AND yearID<=2016 and teamID='ARI' ),

t2 as
(

SELECT *
FROM t1
WHERE siralama=1
GROUP BY yearID,POS),
t3 as

(
select t2.playerID,yearID,POS,siralama,DATE(CONCAT(birthYear,'-',birthMonth,'-',birthDay))  AS DateOfBirth,DATE(CONCAT(yearID,'-',01,'-',01))  AS BegOfYear,TIMESTAMPDIFF(YEAR,DATE(CONCAT(birthYear,'-',birthMonth,'-',birthDay)),DATE(CONCAT(yearID,'-',01,'-',01))) as Age
from t2
inner join Master
on Master.playerID=t2.playerID
ORDER BY yearID)

SELECT yearID,AVG(Age)
FROM t3
GROUP BY yearID;

4 rows affected.


[(2013, Decimal('28.7143')),
 (2014, Decimal('27.2857')),
 (2015, Decimal('27.5714')),
 (2016, Decimal('24.5714'))]

If MLB players typically “peak” at an age of 27 (i.e. they improve up to the age of 27 and regress afterwards), is there any potential correlation between your team’s recent results and its average age?
- No. Our team was doing well in 2013 (rank 16) while the average age was 28.7. If MLB players peak at the age of 27, we should expect the rank improve in 2014 and 2015. But we got worse reselts. 

## Task 2

In [20]:
%%sql
WITH t1 AS
(
SELECT 1.212 AS wOBA_Scale,
       yearID,
       (.691*(BB - IBB) + .721*(HBP) + .878*(H-2B-3B-HR) + 1.242*(2B) + 1.569*(3B) + 2.015*(HR)) / (AB + BB - IBB + SF + HBP) AS League_wOBA
FROM Batting 
)
SELECT  wOBA_Scale,
        SUM(t1.League_wOBA) as sum_League_wOBA
FROM    t1
WHERE   yearID = 2016;

1 rows affected.


/opt/conda/lib/python3.6/site-packages/pymysql/cursors.py:166: Warning: (1365, 'Division by 0')
  result = self._query(query)


[(Decimal('1.212'), 237.7206957332947)]

In [25]:
%%sql
With t1
as
(
SELECT *,RANK() OVER (Partition by POS,yearID ORDER BY InnOuts desc) as siralama
FROM Fielding
WHERE InnOuts != '' AND yearID = 2016 and teamID='ARI' ),

t2 as
(

SELECT *
FROM t1
WHERE siralama=1
GROUP BY yearID,POS),
t3 as

(
select t2.playerID,yearID,POS,siralama,DATE(CONCAT(birthYear,'-',birthMonth,'-',birthDay))  AS DateOfBirth,DATE(CONCAT(yearID,'-',01,'-',01))  AS BegOfYear,TIMESTAMPDIFF(YEAR,DATE(CONCAT(birthYear,'-',birthMonth,'-',birthDay)),DATE(CONCAT(yearID,'-',01,'-',01))) as Age
from t2
inner join Master
on Master.playerID=t2.playerID
ORDER BY yearID)

SELECT t3.playerID,
          POS
                
FROM t3;

InternalError: (pymysql.err.InternalError) (1046, 'No database selected') [SQL: "With t1\nas\n(\nSELECT *,RANK() OVER (Partition by POS,yearID ORDER BY InnOuts desc) as siralama\nFROM Fielding\nWHERE InnOuts != '' AND yearID = 2016 and teamID='ARI' ),\n\nt2 as\n(\n\nSELECT *\nFROM t1\nWHERE siralama=1\nGROUP BY yearID,POS),\nt3 as\n\n(\nselect t2.playerID,yearID,POS,siralama,DATE(CONCAT(birthYear,'-',birthMonth,'-',birthDay))  AS DateOfBirth,DATE(CONCAT(yearID,'-',01,'-',01))  AS BegOfYear,TIMESTAMPDIFF(YEAR,DATE(CONCAT(birthYear,'-',birthMonth,'-',birthDay)),DATE(CONCAT(yearID,'-',01,'-',01))) as Age\nfrom t2\ninner join Master\non Master.playerID=t2.playerID\nORDER BY yearID)\n\nSELECT t3.playerID,\n          POS\n                \nFROM t3;"]

## Task 3

In [18]:
%%sql
WITH T1 AS
(
SELECT *,
                COUNT(*) OVER (PARTITION BY playerID) AS Tot_season,
                CASE
                WHEN teamID = 'FLO' THEN 'MIA'
                WHEN teamID != 'FLO' THEN teamID
                END AS clean_id
FROM     Managers
GROUP BY playerID,
         yearID
HAVING G =162
),
T2 AS 
(
SELECT *
FROM  T1
WHERE T1.Tot_season != 1 AND
      T1.yearID >= 2010  AND
                T1.yearID <= 2016
),
T6 AS
(
SELECT *,
    CASE
                WHEN teamID = 'CHA' THEN 'CHW'
                WHEN teamID = 'CHC' THEN 'CHN'
                WHEN teamID = 'FLO' THEN 'MIA'
                WHEN teamID = 'KCA' THEN 'KCR'
                WHEN teamID = 'LAD' THEN 'LAN'
                WHEN teamID = 'NYA' THEN 'NYY'
                WHEN teamID = 'NYM' THEN 'NYN'
                WHEN teamID = 'SDN' THEN 'SDP'
                WHEN teamID = 'SFG' THEN 'SFN'
                WHEN teamID = 'SLN' THEN 'STL'
                WHEN teamID = 'TBA' THEN 'TBR'
                WHEN teamID = 'WAS' THEN 'WSN'
                ELSE teamID
                END AS clean_id2
FROM Salaries
),
T3 AS
(
SELECT  T2.playerID,
        T2.yearID,
        T2.clean_id,
                T2.W,
                T2.G,
                T2.Tot_season,
        SUM(T6.salary) OVER (PARTITION BY T2.clean_id, T2.yearID) AS Tot_sal
FROM    T2
INNER JOIN T6
ON  T2.clean_id = T6.clean_id2 AND
    T2.yearID = T6.yearID
GROUP BY playerID,
         yearID
),
T4 AS
(
SELECT *,
    T3.W/Tot_sal AS bang
FROM T3
),
T5 AS 
(
SELECT *,
       AVG(T4.bang) OVER (PARTITION BY T4.playerID) AS Avg_bang
FROM  T4

 )
SELECT  T5.playerID,
        T5.yearID,
        Avg_bang
FROM  T5
GROUP BY T5.playerID
ORDER BY Avg_bang DESC
LIMIT 10
;


10 rows affected.


[('actama99', 2010, Decimal('0.00020000')),
 ('leylaji99', 2010, Decimal('0.00020000')),
 ('manueje01', 2010, Decimal('0.00020000')),
 ('gerenbo01', 2010, Decimal('0.00020000')),
 ('gonzafr99', 2011, Decimal('0.00018000')),
 ('halech01', 2015, Decimal('0.00015000')),
 ('francte01', 2010, Decimal('0.00015000')),
 ('millsbr01', 2010, Decimal('0.00015000')),
 ('showabu99', 2011, Decimal('0.00013333')),
 ('farrejo03', 2011, Decimal('0.00013333'))]

In [27]:
%%sql
# How does your team’s coach(es) in 2016 compare to this list?  
# First we will generate the list for 2016

WITH T1 AS
(
SELECT *,
                COUNT(*) OVER (PARTITION BY playerID) AS Tot_season,
                CASE
                WHEN teamID = 'FLO' THEN 'MIA'
                WHEN teamID != 'FLO' THEN teamID
                END AS clean_id
FROM Managers
WHERE teamID = 'ARI'
GROUP BY playerID,
         yearID
HAVING G =162
),
T2 AS 
(
SELECT *
FROM  T1
WHERE T1.Tot_season != 1 AND
      T1.yearID = 2016
),
T6 AS
(
SELECT *,
       CASE
                WHEN teamID = 'CHA' THEN 'CHW'
                WHEN teamID = 'CHC' THEN 'CHN'
                WHEN teamID = 'FLO' THEN 'MIA'
                WHEN teamID = 'KCA' THEN 'KCR'
                WHEN teamID = 'LAD' THEN 'LAN'
                WHEN teamID = 'NYA' THEN 'NYY'
                WHEN teamID = 'NYM' THEN 'NYN'
                WHEN teamID = 'SDN' THEN 'SDP'
                WHEN teamID = 'SFG' THEN 'SFN'
                WHEN teamID = 'SLN' THEN 'STL'
                WHEN teamID = 'TBA' THEN 'TBR'
                WHEN teamID = 'WAS' THEN 'WSN'
                ELSE teamID
                END AS clean_id2
FROM Salaries
),
T3 AS
(
SELECT  T2.playerID,
        T2.yearID,
        T2.clean_id,
        T2.W,
        T2.G,
        T2.Tot_season,
        SUM(T6.salary) OVER (PARTITION BY T2.clean_id, T2.yearID) AS Tot_sal
FROM T2
INNER JOIN T6
ON  T2.clean_id = T6.clean_id2 AND
    T2.yearID = T6.yearID
GROUP BY playerID,
         yearID
),
T4 AS
(
SELECT *,
       T3.W/Tot_sal AS bang
FROM T3
),
T5 AS 
(
SELECT *,
      AVG(T4.bang) OVER (PARTITION BY T4.playerID) AS Avg_bang
FROM  T4

 )
SELECT  T5.playerID,
        T5.yearID,
        Avg_bang
FROM T5
GROUP BY T5.playerID
ORDER BY Avg_bang DESC
LIMIT 10
;
# From the result we can see the coach for ARI is not in TOP 10

UnicodeEncodeError: 'latin-1' codec can't encode character '\u2019' in position 20: ordinal not in range(256)